### PR TITLE
[Core] Add rank-aware CUDA IPC weight sharing primitives

### PR DESCRIPTION
**Pull request by Fyrgo8**
Part of #3064.

## Purpose

This PR introduces the versioned, rank-aware CUDA IPC foundation for sharing provider-owned GPU weight allocations with local consumer processes.

It is intentionally limited to proving the protocol, validation, device mapping, and lifecycle semantics before those primitives are coupled to a production model path. A follow-up PR will integrate one HunyuanImage-3.0 component through the existing model-loading and runtime flows.

### What this PR adds

- A versioned weight manifest covering model/checkpoint/component identity, tensor names, shapes, dtypes, strides, storage offsets and alias relationships, CUDA logical device indices, physical GPU UUIDs, parallel ranks, and parallel world size.
- CUDA IPC handle export and reconstruction through PyTorch multiprocessing reducers.
- Per-consumer reducer export after manifest acknowledgement, so every consumer receives fresh CUDA IPC ref-counter metadata; exports are serialized to protect PyTorch storage mutation.
- An authenticated local provider/consumer handshake that validates the complete manifest before sending or mapping CUDA handles.
- Strict rejection of protocol, model, tensor metadata, device assignment, rank, or parallel-size mismatches.
- Heartbeat-based provider liveness monitoring, bounded liveness checks, graceful consumer shutdown, and prompt listener shutdown.
- Rank-to-device validation based on `(parallel rank, physical GPU UUID)` instead of relying only on process-local CUDA ordinals.

### Scope

This PR does not:

- modify existing model loaders or bypass model-specific `load_weights()` logic
- construct or bind a real model component
- integrate HunyuanImage-3.0 or another model
- add serving modes or runtime orchestration
- support cross-host sharing, provider restart recovery, or dynamic training updates
- support quantization, HSDP, or LoRA

The provider must remain alive while consumers use the mapped CUDA tensors. Model integration, final-weight binding, startup ordering, and full runtime failure propagation remain in the follow-up PR.

## Test Plan

```bash
pytest -s -v tests/distributed/test_ipc_weight_sharing.py \
  -m "core_model and cpu" \
  --run-level=core_model

pytest -s -v tests/distributed/test_ipc_weight_sharing_cuda.py \
  -m "core_model and cuda" \
  --run-level=core_model

ruff check \
  vllm_omni/distributed/ipc_weight_sharing.py \
  vllm_omni/distributed/__init__.py \
  tests/distributed/test_ipc_weight_sharing.py \
  tests/distributed/test_ipc_weight_sharing_cuda.py
```

**vLLM Version:** 0.25.0

**vLLM-Omni Commit:** `eaed7e12`

## Test Result

Current revision:

- CPU protocol and lifecycle tests: `18 passed`
- Concurrent protocol stress: 20 consumers received 20 unique exported handle sets
- Ruff lint and format checks: passed
- Python compile check: passed
- A two-consumer CUDA regression now asserts that both processes map the same provider allocation while receiving distinct `(ref_counter_handle, ref_counter_offset)` pairs
- The new CUDA regression has not been rerun yet because the GPU test host was unavailable when this update was prepared

Previously validated CUDA baseline:

- 2 x NVIDIA GeForce RTX 5090, 32 GB
- NVIDIA driver 595.71.05
- CUDA Toolkit 13.0
- PyTorch 2.11.0+cu130
- Python 3.12.3
- CUDA IPC tests: `2 passed`
- CUDA `sm_120` compilation check: passed
- Consumer allocator delta after mapping a 16 MiB provider allocation: `0 bytes`
- Provider in-place updates were visible in the consumer
- Two-GPU rank/device UUID mapping passed
- Lifecycle stress checks completed with zero false provider-exit callbacks and zero remaining weight-sharing threads
- GPU memory returned to the 2 MiB idle baseline on both devices after the tests

### CODE REVIEW

**Comment by Fyrgo8** [comment]
@SamitHuang PR 1 for #3064 is ready for review. This PR is limited to the versioned sharing protocol, manifest and rank/device validation, per-consumer CUDA IPC handle export, and provider/consumer lifecycle handling. Model-loader and HunyuanImage-3.0 integration remain out of scope for the follow-up PR 2.

File: vllm_omni/distributed/ipc_weight_sharing.py [hunk 1]
```diff
@@ -0,0 +1,1139 @@
...
+        self._serve_thread = Thread(target=self.serve_forever, name="weight-sharing-provider", daemon=True)
+        self._serve_thread.start()
+        return self._serve_thread
+
+    def serve_forever(self) -> None:
+        """Accept compatible consumers until :meth:`stop` is called."""
+        while not self._stop_event.is_set():
+            try:
+                connection = self._listener.accept()
+            except (OSError, EOFError):
```
Comment (SamitHuang) [inline vllm_omni/distributed/ipc_weight_sharing.py:761]: `Listener.accept()` with an `authkey` raises `multiprocessing.AuthenticationError` (not `OSError`) when the peer does not complete the HMAC handshake. `_wake_listener()` connects with a raw `socket.socket`, so `stop()` can either leave `serve_forever` blocked on `accept()` or kill the listener thread with an uncaught `AuthenticationError`. Catch `AuthenticationError` here (and continue the loop when `_stop_event` is set), or wake the listener with an authenticated `Client` instead of a raw TCP/UNIX connect. Add a regression around `test_provider_stop_unblocks_listener_thread` that fails if the serve thread is still alive after `stop()`.

File: vllm_omni/distributed/ipc_weight_sharing.py [hunk 2]
```diff
@@ -0,0 +1,1139 @@
...
+        self.tensors = dict(tensors)
+        self._handle_exporter = handle_exporter
+        self.handshake_timeout = handshake_timeout
+        self.heartbeat_interval = heartbeat_interval
+        self._stop_event = Event()
+        self._connections: set[Connection] = set()
+        self._threads: set[Thread] = set()
+        self._connections_lock = Lock()
+        self._handle_export_lock = Lock()
+        self._listener = Listener(endpoint.address, family=endpoint.family, authkey=endpoint.authkey)
```
Comment (SamitHuang) [inline vllm_omni/distributed/ipc_weight_sharing.py:718]: `Listener` is bound in `__init__` with the process umask and no cleanup of a leftover path. A crashed provider leaves `/tmp/vllm_shared_*_rankN.sock` behind so the next start fails with address-in-use, and a world-readable socket plus pickle-based handle transfer is enough for any local process that obtains the authkey to map CUDA weights. Unlink a stale socket before bind, `chmod` the path to `0o600` after listen, and give `WeightSharingProvider` a context manager so the fd/path cannot leak if `start()` is never paired with `stop()`. There is also no AF_UNIX test covering bind, stale-path recovery, or permissions.

File: tests/distributed/test_ipc_weight_sharing_cuda.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+            if process.is_alive():
+                process.kill()
+        for command_queue, result_queue in zip(command_queues, result_queues, strict=True):
+            command_queue.close()
+            result_queue.close()
+
+    assert all(process.exitcode == 0 for process in processes)
+
+
+@hardware_test(res={"cuda": "L4"}, num_cards=2)
```
Comment (SamitHuang) [inline tests/distributed/test_ipc_weight_sharing_cuda.py:250]: `@hardware_test(..., num_cards=2)` plus `device_count() < 2` skip means both rank/UUID tests are no-ops on the CUDA ready job, which uses `mirror_hardwares: l4_1` and `-m 'core_model and cuda and L4'`. The feature this PR exists to prove — pairing by `(parallel_rank, physical GPU UUID)` rather than process-local ordinals — therefore has no CI signal. Move these two tests onto a 2-GPU shard (or add a dedicated `distributed_cuda` step) and keep a 1-GPU IPC mapping test on `l4_1`.

File: tests/distributed/test_ipc_weight_sharing_cuda.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+        args=(
+            (provider.endpoint.address, provider.endpoint.family, provider.endpoint.authkey),
+            manifest.to_payload(),
+            command_queue,
+            result_queue,
+            0,
+        ),
+    )
+    previous_visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES")
+    os.environ["CUDA_VISIBLE_DEVICES"] = str(provider_device_index)
```
Comment (SamitHuang) [inline tests/distributed/test_ipc_weight_sharing_cuda.py:291]: CUDA is already initialized in the parent (tensor created on `cuda:1`) before this writes `CUDA_VISIBLE_DEVICES`. Changing that env after init does not remap the parent's devices and is a known source of cross-test contamination; spawn children should get the mask via `Process(..., env=...)` or an explicit child wrapper, not by mutating the parent. Also, the child is started with `consumer_device_index=0` while `CUDA_VISIBLE_DEVICES` is a single GPU, so this does not exercise `remap_cuda_handle_device` against a provider handle that still contains `cuda:1` under a shared visibility mask. Reconstruct the remap with a shared `CUDA_VISIBLE_DEVICES` list and different logical ordinals, or drop the parent env mutation entirely.

File: tests/distributed/test_ipc_weight_sharing.py [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+        provider.stop()
+
+
+def test_provider_exports_fresh_handles_for_each_consumer() -> None:
+    manifest = _manifest()
+    export_count = 0
+
+    def export_handles(
+        tensors: Mapping[str, object],
+        current_manifest: WeightManifest,
```
Comment (SamitHuang) [inline tests/distributed/test_ipc_weight_sharing.py:420]: This only checks two sequential consumers. The PR description claims a 20-consumer stress where every exported handle set is unique, but that test is not in the tree, and all CPU coverage uses `WeightSharingEndpoint.tcp` with fake exporters. Add the concurrent uniqueness test against the real handshake (threaded consumers, unique `(handle, ref_counter)` tuples) and at least one AF_UNIX happy-path plus stale-socket restart case, otherwise the lifecycle claims are not gated by CI.

**Review by SamitHuang — CHANGES_REQUESTED** [review]
Review of the rank-aware CUDA IPC primitives for #3064. The protocol/manifest split is the right scope for PR 1, but a few lifecycle and coverage holes will bite once this is wired into HunyuanImage-3.0.

Blocking:
1. `stop()` can fail to unblock the listener: `_wake_listener()` uses a raw TCP/UNIX connect while `Listener.accept()` expects the multiprocessing auth handshake, and `AuthenticationError` is not caught.
2. AF_UNIX sockets are neither chmod'd nor rebound over a stale path, and there is no Unix-socket test despite the RFC using `/tmp/..._rankN.sock`.
3. The two-GPU rank/UUID tests never run in current CUDA `test-ready.yml` (`Distributed Test` is `l4_1`), and the ordinal-remap test mutates the parent's `CUDA_VISIBLE_DEVICES` after CUDA is already initialized.

The PR body also states a 20-consumer uniqueness stress and that the new two-consumer CUDA regression has not been rerun on GPU. Neither claim is backed by committed tests plus a current L4 result. Please land those before merge; do not treat the RTX 5090 notes as a substitute for the L4 CI job.

File: vllm_omni/distributed/ipc_weight_sharing.py [hunk 6]
```diff
@@ -0,0 +1,1139 @@
...
+
+def remap_cuda_handle_device(handle: TensorHandle, device_index: int) -> TensorHandle:
+    """Rewrite a PyTorch CUDA reducer handle for a consumer-local ordinal."""
+    if device_index < 0:
+        raise ValueError("CUDA device index must be non-negative")
+
+    import torch
+
+    device_positions = [
+        index for index, value in enumerate(handle) if isinstance(value, torch.device) and value.type == "cuda"
```
Comment (lishunyang12) [inline vllm_omni/distributed/ipc_weight_sharing.py:527]: Replace handle[6] with the integer device_index; PyTorch's CUDA reducer doesn't store a torch.device.

File: vllm_omni/distributed/ipc_weight_sharing.py [hunk 7]
```diff
@@ -0,0 +1,1139 @@
...
+        heartbeat_interval: float = 1.0,
+        on_provider_exit: ProviderExitCallback | None = None,
+    ) -> MappedWeights:
+        if handshake_timeout <= 0:
+            raise ValueError("handshake_timeout must be positive")
+        if heartbeat_interval <= 0:
+            raise ValueError("heartbeat_interval must be positive")
+        if validate_local_device:
+            _validate_local_manifest_device(expected_manifest)
+        connection = Client(endpoint.address, family=endpoint.family, authkey=endpoint.authkey)
```
Comment (lishunyang12) [inline vllm_omni/distributed/ipc_weight_sharing.py:1063]: Apply handshake_timeout to connection and authentication; Client() currently blocks indefinitely.
